### PR TITLE
Expire local tracks

### PR DESCRIPTION
**Pull request by zanderz**
@maxtaco, This mainly makes sure locally stored tracks get aged out instead of lasting forever, but I'm unsure whether this covers the intent of 2447. Would appreciate at least a first look at this draft.

### CODE REVIEW

File: go/libkb/env.go [hunk 1]
```diff
@@ -545,6 +546,13 @@ func (e *Env) GetScraperTimeout() time.Duration {
 	)
 }
 
+func (e *Env) GetLocalTrackMaxAge() time.Duration {
+	return e.GetDuration(LocalTrackMaxAge,
+		func() (time.Duration, bool) { return e.cmd.GetLocalTrackMaxAge() },
+		func() (time.Duration, bool) { return e.getEnvDuration("KEYBASE_LOCAL_TRACK_MAX_AGE") },
+		func() (time.Duration, bool) { return e.config.GetLocalTrackMaxAge() },
```
Comment (zanderz) [inline go/libkb/env.go]: I just copied some other global, configurable variable

File: go/libkb/track.go [hunk 2]
```diff
@@ -375,6 +378,19 @@ func LocalTrackChainLinkFor(tracker, trackee keybase1.UID, g *GlobalContext) (re
 		g.Log.Debug("| unpack failed -> %s", err)
 		return
 	}
+
+	linkEtime := cl.GetCTime().Add(g.Env.GetLocalTrackMaxAge())
```
Comment (zanderz) [inline go/libkb/track.go]: Also not positive the existing creation time of the track is the right thing to use

**Comment by maxtaco** [comment]
This is a start, but there's a lot missing here.  There are several cases to consider:

* We still want regular local tracking to work and not to expire.
* Expiring local tracking is only for when some of the proofs are broken
* The command-line experience needs to change in the case of tracking someone who has broken proofs. You should have the ability to drop into one of three outcomes: remote track with broken proofs, permanent local track with broken proofs, or an expiring local track.  If all proofs work, no need to show the option for a local expiring track.
* The electron UI needs to access to these sorts of features, likely we'll have to change `TrackOptions` to express it.

I'm not fully sure if we can get away with storing the temporary and permanent local tracking links in the same slot in level DB.  It might be simplest to store them in two different locations.

File: go/engine/track_token_test.go [hunk 3]
```diff
@@ -76,3 +78,97 @@ func TestTrackTokenIdentify2(t *testing.T) {
...
+
+	// Identify should work once more because we signed with failures
+	eng = NewResolveThenIdentify2(tc.G, arg)
+	eng.testArgs = &Identify2WithUIDTestArgs{noCache: true}
+	var err error
+	// Should not get an error
+	if err = RunEngine(eng, ctx); err != nil {
+		t.Logf("Identify failure: %v", err)
+		// TODO: The engine is saying there is 1 failure still.
+		//t.Fatal("Expected to pass identify")
```
Comment (zanderz) [inline go/engine/track_token_test.go]: This! Does not seem right but I'm not sure what to do about it.

**Comment by zanderz** [comment]
Thanks for all the further info, but I have been spinning my wheels on this and it still doesn't seem quite right. Possibly bit off more than I could chew - or maybe it's close and just another tweak or two?

**Comment by maxtaco** [comment]
Seems like good progress. What seems not quite right?

**Comment by zanderz** [comment]
In the test (copied from the soft fail test), Identify2 fails after the locally stored track has been read back out. I thought that if the rooter check fails and the track is stored locally that way, Identify2 should use it next time - or maybe that's not the case? It's unclear when this local, temporary track will come to bear and save the user another interaction before it expires. My test doesn't quite capture that.

File: go/engine/track_token_test.go [hunk 4]
```diff
@@ -76,3 +78,97 @@ func TestTrackTokenIdentify2(t *testing.T) {
...
+
+	// Track tracy
+	teng := NewTrackToken(&targ, tc.G)
+	if err := RunEngine(teng, ctx); err != nil {
+		t.Fatal(err)
+	}
+
+	// Now make her Rooter proof fail
+	t.Logf("-- Set for Hard Fail --")
+	flakeyAPI.hardFail = true
```
Comment (maxtaco) [inline go/engine/track_token_test.go]: Why is this a hard failure? This mechanism (local caching/snoozing for temporary failures) really only ought to work for "soft" failures.

**Comment by maxtaco** [comment]
BTW, we need another test to make sure this feature works with permanent local tracking.

**Comment by maxtaco** [comment]
The high-level points are:

1.your test was broken, since it was using hard proof check failures, which should never be snoozed; I fixed the test for you so I could get to bug 2 which is....
1. there are bugs in either your storing, or loading, or using of your temporary local tracks that you should look into; I didn't get a chance to dig any deeper

I recommend putting in some more Debug/Vlog1 code to track down what local tracking statement you're writing to LevelDB, and what statement you're reading back out.  Also, you should make your test more robust and it should query the results of identify to assert that the right tracking components failed, as I do in my snooze test.

**Comment by maxtaco** [comment]
See my comments in ff7cef6 for what was wrong with your test...

**Comment by zanderz** [comment]
Hard proof failures: I thought all along this was for hard failures; we discussed the case where the site hosting the proof goes down, as opposed to just rate limiting. Isn't that a hard failure? Anyway, good to know, thanks again, I'll figure out the reading/writing.

**Comment by maxtaco** [comment]
No no no definitely only for soft errors.  Soft errors are of the form: 429 rate limits, or 503 server not available, etc.  The mental model is that GitHub went down, and now all of your tracker statements are failing.  So GitHub is sending you a 503, and you just want put off any decision for a day or so.

**Comment by zanderz** [comment]
`track.go:386: [D] | := local track created 2016-02-11 13:23:03 -0800 PST, expires: 2016-02-13 13:23:03 -0800 PST, it is now 1984-04-06 00:01:00 +0000 UTC`
I at first thought this creation date was from the server but probably it is just using time.Now() instead of the global context clock.

**Comment by maxtaco** [comment]
Right, that's why I recommended not resetting the fake clock to now, to make these issues more obvious.  Almost certainly you are correct, most of the code uses `time.Now()`, but we've migrated over to using the clock dangling off of the global context as we've needed it.

**Comment by maxtaco** [comment]
Seems like this PR is currently failing test...

**Comment by maxtaco** [comment]
Also, how did you test the CLI interaction? I want to try to walk through it myself.

**Comment by maxtaco** [comment]
As for the CLI UX, here is a potential design:

### Setup

alice runs `keybase track bob`

### New UX Behavior:

If alice previously tracked bob, but one of the proofs that alice previously tracked is now soft-failing, alice is prompted:

* Some of previously tracked proofs are failing. **I**gnore for 24 hours, **A**ccept these new changes or **C**ancel? [I/a/c]

If the User answers **A**ccept, then follow up:

* Publish these changes publicly to keybase.io? [Y/n]

**Comment by malgorithms** [comment]
I would bracket the first letters (not just colorify), just to make it clearer: [I]gnore for 24 hours, [A]ccept these new changes or [C]ancel? [I/a/c]

**Comment by maxtaco** [comment]
To close to "Abort / Retry / Fail" ?

**Comment by maxtaco** [comment]
Anyways, as for testing, I was thinking:

* Dial down expiration time to ~10s, so we can test they expire properly
* Put something like `github.com 5.5.5.5` in `/etc/hosts.conf` or the windows equivalent. That should give you a soft failure.

**Comment by zanderz** [comment]
OK so I squashed/rebased onto current master and force-pushed this branch, hope that's ok!

**Comment by maxtaco** [comment]
Took a brief skim --- looking good, but I will go over it tomorrow.

BTW, what did you decide about outputting more information in the case of an expiring local track masking a failure?

**Comment by zanderz** [comment]
I decided not to do it right now! What I did do was attempt to make the
letters bold with ColorString(), etc., and figured out those prompts are
rendered in miniterm, which didn't seem to do it right, at least on windows
- something about the square brackets and the color codes maybe? I took
that back out but that's as far as I got tonight.

**Comment by maxtaco** [comment]
I'm putting in some UX for saying it's a temporary track. Good thing I did!

```
You last locally tracked t_charlie on 2016-02-17 09:38:10 EST (track expires on 2032-02-13 09:38:10 EST)
```

WAT?  

Looking into it...

**Comment by maxtaco** [comment]
I fixed the UI issue we spoke of in commit 8a00288 above.  There's a big problem though.

The problem is this:
* I track alice, everything works
* I track alice, github fails, but I commit an expiring local track
* I track alice, github is back up, but my expiring local track *is still valid*

I now get a crazy experience, that's asking me to track something that appears to work.  I'm not really sure what to do about this problem, thinking about it... Thoughts?

One proposal is to load the local track and the permanent track into the identify state, and to try the permanent track first, and fallback to the temporary.  If the permanent track works without a problem, then use that, otherwise use the expiring local track.  And update the UI accordingly. This might involve a lot of disruption though to the identify logic.

**Comment by malgorithms** [comment]
> One proposal is to load the local track and the permanent track into the identify state, ...

From the user experience perspective, that seems to produce the right result. I think from a user's perspective as long as the state of the user satisfies *either* the expiring local track or the full track, then the user shouldn't be bothered.

**Comment by maxtaco** [comment]
There is some last-minute messaging --- if you didn't need the local track at all, the UX shouldn't show that it was used.

**Comment by zanderz** [comment]
I suppose that comparing each individual proof result between expiring and
permanent tracks isn't any simpler than what you've proposed.

**Comment by maxtaco** [comment]
That sounds like what I've proposed.  The order of operations would be to favor the permanent over the expiring, but to fall back to the expiring.  We keep a bit of the form "was the expiring track used" and bubble that back up to the CLI

**Comment by maxtaco** [comment]
(you couldn't do the whole operation twice, since that sends all sorts of UI RPC side-effects)

**Comment by maxtaco** [comment]
nit: can you make the input accept upper or lower case?

**Comment by zanderz** [comment]
Sure

**Comment by zanderz** [comment]
Will this be acceptable in the hint?
Some previously tracked proofs are failing. [I]gnore for 1m0s, [A]ccept
these changes or [C]ancel? (I, i, A, a, C, c):

**Comment by maxtaco** [comment]
hmmm, that is pretty ugly.  i'd say the hint should stay the same, but the user can user either?

**Comment by maxtaco** [comment]
Still missing:

* New test for this new case (mentioned above)
* To say in the header when the local temporary track expires
* To clean up potentially `GetDiff`

**Comment by maxtaco** [comment]
Remaining issues:

* Input of `c` doesn't cancel like `C` does at prompt
* Still need to display when temporary tracks will expire.

**Comment by maxtaco** [comment]
Well at least C/I is passing!

**Comment by maxtaco** [comment]
All systems go with my latest commits!

**Comment by maxtaco** [comment]
@zanderz do you want to take a read over them, and then we can merge?

**Comment by zanderz** [comment]
Will do.

**Comment by zanderz** [comment]
Everything seems to work, LGTM. Got a question or 2:
1. don't I still need to make a test for the case where the proof stops
failing before the temporary track expires?
2. Are we sure the temp local track gets deleted after it expires?

**Comment by maxtaco** [comment]
Never tested (2).  (1) is good, you're right, if you can crank it out...

**Pull request merged**